### PR TITLE
New integration test 

### PR DESCRIPTION
**Pull request by harvey-intel**
* Add unit test for metric computation
* Update integration test coverage

### CODE REVIEW

File: integration_test/test_xgboost.py [hunk 1]
```diff
@@ -54,28 +123,52 @@ def _verify_parameter_logging(self, run, params):
...
+    assert run.metadata['Dataset columns'] == self.run_params['dtrain'].num_col()
+    assert run.metadata['Dataset rows'] == self.run_params['dtrain'].num_row()
+    assert run.metadata['Number of Test Sets'] == len(self.run_params['evals'])
     assert run.metadata['Python Version'] == platform.python_version()
     assert run.metadata['XGBoost Version'] == xgb.__version__
-    self._verify_parameter_logging(run, xgb_params['params'])
-    assert run.assignments['num_boost_round'] == xgb_params['num_boost_round']
+
+  def test_run(self):
+    self._form_random_run_params()
```
Comment (yuhong-sigopt) [inline integration_test/test_xgboost.py:154]: I prefer for each type of model, add a new test case. In each case, there should be several random samples.

File: integration_test/test_xgboost.py [hunk 2]
```diff
@@ -1,51 +1,120 @@
...
+      'eval_metric': ['logloss', 'aucpr', 'error'],
+    }
+  elif task == 'multiclass':
+    return {
+      'objective': random.choice(['multi:softmax', 'multi:softprob']),
+      'eval_metric': ['mlogloss', 'merror'],
+    }
+  else:
+    return {
+      'objective': random.choice(['reg:squarederror']),
```
Comment (ericlee0803) [inline integration_test/test_xgboost.py:81]: If we're considering all the learning objectives for binary and multi class classification, might as well do the same for regression: 

`['reg:squarederror', 'reg:squaredlogerror', 'reg:logistic', 'reg:pseudohubererror', 'reg:gamma', 'reg:tweedie']`

(not sure if all of these are compatible with standard floating point data)

File: test/xgboost/test_compute_metric.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    assert report['0']['f1-score'] == 0
+    assert report['weighted avg']['precision'] == 0
+    assert report['weighted avg']['recall'] == 0
+    assert report['weighted avg']['f1-score'] == 0
+
+  def test_binary_classification_one_pred_label(self):
+    y_true = numpy.array([1, 1, 1, 0, 0, 0, 0, 0, 0, 0])
+    y_pred = numpy.ones(10, dtype=int)
+    y_true[:3] = 1
+    # y_true = [1, 1, 1, 0, 0, 0, 0, 0, 0, 0]
```
Comment (ericlee0803) [inline test/xgboost/test_compute_metric.py:47]: delete?

File: test/xgboost/test_compute_metric.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    y_true = 2 * numpy.ones(n_samples)
+    assert numpy.isclose(compute_mae(y_true, numpy.zeros(n_samples)), 2.0)
+    assert numpy.isclose(compute_mse(y_true, numpy.zeros(n_samples)), 4.0)
+
+    y_true = numpy.random.randn(n_samples)
+    assert numpy.isclose(compute_mae(y_true, y_true), 0)
+    assert numpy.isclose(compute_mse(y_true, y_true), 0)
+
+    y_pred = numpy.random.randn(n_samples)
+    assert numpy.isclose(compute_mae(y_true, y_pred), compute_mae(y_pred, y_true))
```
Comment (ericlee0803) [inline test/xgboost/test_compute_metric.py:88]: compute_mae/mse is used twice here... is one call supposed to be the sklearn mean_absolute_error / mean_squared_error functions?

**Review by ericlee0803 — CHANGES_REQUESTED** [review]

File: test/xgboost/test_compute_metric.py (hunk 4 above)
Comment (ericlee0803) [reply test/xgboost/test_compute_metric.py:88]: Ah I see you are checking symmetry in the arguments. At any rate, would be nice to test against sklearn MSE and MAE functions

File: integration_test/test_xgboost.py (hunk 1 above)
Comment (harvey-intel) [reply integration_test/test_xgboost.py:154]: I see what you are saying and I agree in general. However, I want to wait till I know how long this test takes on circleci before deciding. If each test (xgb.train) takes <30 seconds, then sure let's split this into one test for each task. 

Mostly, I just want to avoid having people waiting for this test to finish every time.

File: integration_test/test_xgboost.py (hunk 2 above)
Comment (harvey-intel) [reply integration_test/test_xgboost.py:81]: The problem with `squaredlogerror` or `logistic` error is that they require the y to be of certain value ranges. Didn't have a quick way of dealing with it using the `make_regression` function. 

The rest should be ok? I guess

File: integration_test/test_xgboost.py (hunk 1 above)
Comment (yuhong-sigopt) [reply integration_test/test_xgboost.py:154]: I think one small model for each model type is OK.

Comment (harvey-intel) [reply integration_test/test_xgboost.py:154]: Cool. I'll make the change.

**Review by yuhong-sigopt — APPROVED** [review]

File: integration_test/test_xgboost.py (hunk 2 above)
Comment (harvey-intel) [reply integration_test/test_xgboost.py]: hmm `gamma` and `tweedie` all requires y > 0.

Comment (ericlee0803) [reply integration_test/test_xgboost.py]: hmm okay never mind just reg:squarederror is fine then

**Review by ericlee0803 — APPROVED** [review]

**Comment by harvey-intel** [comment]
Have a few things that I need iron out after rebasing on SysMetadata

File: sigopt/xgboost/run.py [hunk 5]
```diff
@@ -157,7 +157,12 @@ def check_learning_task(self):
 
   def log_feature_importances(self, importance_type='weight', fmap=''):
     scores = self.bst.get_score(importance_type=importance_type, fmap=fmap)
-    scores = dict(sorted(scores.items(), key=lambda x:(x[1], x[0]), reverse=True)[:FEATURE_IMPORTANCES_MAX_NUM_FEATURE])
+    # NOTE: do not log importances if there is no split at all.
+    if not scores:
+      return
```
Comment (harvey-intel) [inline sigopt/xgboost/run.py:162]: cc @yuhong-sigopt @ericlee0803 , I decided to not log feature importances if none of the tree has a split. Let me know if you are ok with this.

Comment (ericlee0803) [reply sigopt/xgboost/run.py:162]: Fine w/ me

Comment (yuhong-sigopt) [reply sigopt/xgboost/run.py:162]: I am ok.
